src/esp/gfx/ssao/HBAOHelper2.cpp

```diff
@@ -0,0 +1,784 @@
+// Copyright (c) Meta Platforms, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "HBAOHelper2.h"
+#include <Corrade/Utility/Resource.h>
+#include <Magnum/GL/AbstractShaderProgram.h>
+#include <Magnum/GL/Buffer.h>
+#include <Magnum/GL/Context.h>
+#include <Magnum/GL/Framebuffer.h>
+#include <Magnum/GL/Mesh.h>
+#include <Magnum/GL/Shader.h>
+#include <Magnum/GL/Texture.h>
+#include <Magnum/GL/TextureArray.h>
+#include <Magnum/GL/TextureFormat.h>
+#include <Magnum/GL/Version.h>
+#include <Magnum/ImageView.h>
+#include <Magnum/Magnum.h>
+#include <Magnum/Math/Matrix4.h>
+#include <Magnum/Math/Vector4.h>
+#include <Magnum/PixelFormat.h>
+
+#include <random>
+
+// #include "src/shaders/ssao/common.h" // can't access this from here
+
+namespace Mn = Magnum;
+namespace Cr = Corrade;
+
+namespace esp {
+namespace gfx {
+namespace ssao {
+
+namespace {
+
+// sloppy: duplicate these from common.h
+#define AO_RANDOMTEX_SIZE 4
+static const int MAX_SAMPLES = 8;
+static const int HBAO_RANDOM_SIZE = AO_RANDOMTEX_SIZE;
+static const int HBAO_RANDOM_ELEMENTS = HBAO_RANDOM_SIZE * HBAO_RANDOM_SIZE;
+
+struct SceneData {
+  Mn::Matrix4 viewProjMatrix;
+  Mn::Matrix4 viewMatrix;
+  Mn::Matrix4 viewMatrixIT;
+
+  Mn::Vector2ui viewport;
+  Mn::Vector2ui _pad;
+};
+
+struct HBAOData {
+  float RadiusToScreen;  // radius
+  float R2;              // 1/radius
+  float NegInvR2;        // radius * radius
+  float NDotVBias;
+
+  Mn::Vector2 InvFullResolution;
+  Mn::Vector2 InvQuarterResolution;
+
+  float AOMultiplier;
+  float PowExponent;
+  Mn::Vector2 _pad0;
+
+  Mn::Vector4 projInfo;
+  Mn::Vector2 projScale;
+  int projOrtho;
+  int _pad1;
+
+  Mn::Vector4 float2Offsets[AO_RANDOMTEX_SIZE * AO_RANDOMTEX_SIZE];
+  Mn::Vector4 jitters[AO_RANDOMTEX_SIZE * AO_RANDOMTEX_SIZE];
+};
+
+// class BilateralBlurShader: Mn::GL::AbstractShaderProgram {
+// public:
+//   BilateralBlurShader() {
+//     Cr::Utility::Resource rs{"default-shaders"};
+
+//     Mn::GL::Shader vert{glVersion, Mn::GL::Shader::Type::Vertex};
+//     Mn::GL::Shader frag{glVersion, Mn::GL::Shader::Type::Fragment};
+
+//     vert.addSource(rs.getString("ssao/common.h"));
+//     vert.addSource(rs.getString("ssao/fullscreenquad.vert.glsl"));
+//     frag.addSource(rs.getString("ssao/bilateralblur.frag.glsl"));
+
+//     CORRADE_INTERNAL_ASSERT(vert.compile() && frag.compile());
+
+//     attachShaders({vert, frag});
+//     CORRADE_INTERNAL_ASSERT(link());
+//   }
+// };
+
+class HBAOShader : public Mn::GL::AbstractShaderProgram {
```

Comment:
If we find this class generally useful (as a way to port external OpenGL code), we could move it out of this file and rename it `PublicShaderProgram` (similar to an earlier suggestion from @mosra).